### PR TITLE
Refactor uplift trees by replacing dict with list

### PR DESCRIPTION
**Pull request by jeongyoonlee**
## Proposed changes

This PR refactors `UpliftTreeClassifier` and `UpliftRandomForestClassifier` in `inference.tree`. 

Updates include:
- Replacing `dict` with `list` for optimization with `numba` or `cython` (#54)
- Updating docstrings to reflect the correct input and return types (#227)
- Using `np.random.RandomState` in `UpliftTreeClassifier` and `UpliftRandomForestClassifier` (#401)
- Updating `UpliftRandomForestClassifier.classes_` to include `control_name` as the first item
- Moving generic functions (`kl_divergence`, `entropyH`) out of the `UpliftTreeClassifier` class to `inference.tree.utils`
- Removing unused/unnecessary input/return arguments
- Updating tests and example notebooks accordingly

## Types of changes

What types of changes does your code introduce to **CausalML**?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/uber/causalml/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/uber/causalml)
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

No change in public APIs except `UpliftRandomForest.classes_`, which now includes `control_name` as the first item.

### CODE REVIEW

File: causalml/inference/tree/models.py [hunk 1]
```diff
@@ -920,50 +822,44 @@ def growDecisionTreeFrom(self, X, treatment, y, evaluationFunction, max_depth=10
...
         else:
-            pt = currentNodeSummary[suboptTreatment][0]
-            nt = currentNodeSummary[suboptTreatment][1]
-            pc = currentNodeSummary[self.control_name][0]
-            nc = currentNodeSummary[self.control_name][1]
-            p_value = (1. - stats.norm.cdf((pc - pt) / np.sqrt(pt * (1 - pt) / nt + pc * (1 - pc) / nc))) * 2
+            p_t = currentNodeSummary[suboptTreatment][0]
+            n_t = currentNodeSummary[suboptTreatment][1]
+
+        p_value = (1. - stats.norm.cdf((p_c - p_t) / np.sqrt(p_t * (1 - p_t) / n_t + p_c * (1 - p_c) / n_c))) * 2
```
Comment (paullo0106) [inline causalml/inference/tree/models.py:862]: to confirm, do we need to use `p_t - p_c` when `maxDiff > 0` (if I understand the origin logic correctly) ?

File: causalml/inference/tree/models.py [hunk 2]
```diff
@@ -1008,47 +904,36 @@ def growDecisionTreeFrom(self, X, treatment, y, evaluationFunction, max_depth=10
...
-                elif evaluationFunction == self.evaluate_DDP:
-                    if self.control_name in leftNodeSummary and self.control_name in rightNodeSummary:
-                        leftScore1 = evaluationFunction(leftNodeSummary, control_name=self.control_name)
-                        rightScore2 = evaluationFunction(rightNodeSummary, control_name=self.control_name)
-                        gain = np.abs(leftScore1 - rightScore2)
-                        gain_for_imp = np.abs(len(X_l) * leftScore1 - len(X_r) * rightScore2)
-                    else:
-                        gain = 0
+                elif self.evaluationFunction == self.evaluate_DDP:
+                    leftScore1 = self.evaluationFunction(leftNodeSummary, control_name=self.control_name)
```
Comment (paullo0106) [inline causalml/inference/tree/models.py:920]: `control_name` parameter seems to be removed in `evaluate_DDP()`, we then need to change these two lines accordingly

**Review by paullo0106 — COMMENTED** [review]
Looks really nice and helpful!

File: causalml/inference/tree/models.py (hunk 1 above)
Comment (jeongyoonlee) [reply causalml/inference/tree/models.py:869]: I think you're right. @t-tte, could you confirm?

File: causalml/inference/tree/models.py (hunk 2 above)
Comment (jeongyoonlee) [reply causalml/inference/tree/models.py]: Done. Thanks for catching this!

File: causalml/inference/tree/models.py (hunk 1 above)
Comment (jeongyoonlee) [reply causalml/inference/tree/models.py]: Fixed it.

**Review by paullo0106 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**